# TAREA FINAL: Datos de vivienda
### Iker Aguirre Otaegui

<p> El negocio inmobiliario es uno de los motores de la economía de España, el carácter turístico de nuesta península hace que sean muchos los inversores extranjeros que se decidan por buscar una inversión de bajo riesgos en nuestras ciudades costeras o grandes urbes. 
Muchas de estas inversiones se hacen sobre grandes bolsas de inmuebles que deben ser analizados previamente para comprobar la rentabilidad del porfolio </p>

<!-- <p> En este caso vamos a trabajar con una tabla que contienen información de distintos inmuebles repart

Todas las cuestiones se deben realizar sobre el conjunto de casos que representen viviendas ubicadas en zonas residenciales **(alta, media y baja densidad)**

**MUY IMPORTANTE:** En las otras prácticas he detectado colaboraciones involucrando varias personas y he sido flexible aunque a algunos os lo he mencionado en las correcciones, porque al final el trabajo de analista de datos es un trabajo colaborativo. Sin embargo, este trabajo es individual, así que cuidado con las colaboraciones.

Las variables de las que se compone el dataset son:

|NOMBRE VARIABLE|DESCRIPTOR|VALORES|
| --- | --- | --- |
|Order|Variable de identificación|1 a 2930|
|MS Zoning|Zona de ubicación de la vivienda|"A rural, C comercial, FV residencial flotante, I industrial, RH residencial alta densidad, RL residencial baja densidad, RM residencial media densidad"|
|Lot Frontage|Longitud de la fachada en pies||
|Lot Area|Superficie de la vivienda en pies cuadrados||
|Land Contour|Contorno del terreno circundante|"Lvl llano, Bnk Tipo bancal, HLS Ladera, Low Depresión"|
|Land Slope|Tipo de pendiente de la vivienda|" Gtl pendiente suave, Mod pendiente moderada, Sev fuerte pendiente"|
|Overall Qual|Grado de calidad de materiales y acabado de la vivienda|De 1 (Muy pobre) a 10 (Excelente)|
|Year Built|Año de construccion de la vivienda||
|Year Remod/Add|Año de última reforma de la vivienda||
|Mas Vnr Type|Tipo de revestimiento exterior|" BrkCmn Ladrillo normal, BrkFace Ladrillo visto, CBlock Bloque de cemento, None Ninguna, Stone Piedra "|
|Exter Qual|Calidad de revestimiento exterior|"Ex Excelente,Gd Bueno,TA Media,Fa Justo"|
|Bsmt Cond|Estado general del sótano|"Ex Excelente, Gd Bueno, TA Media, Fa Justo, Po Pobre,Ss sin sótano"|
|Total Bsmt SF|Superficie del sótano en pies cuadrados|
|Heating QC|Calidad de la calefacción|"Ex Excelente,Gd Bueno,TA Media,Fa Justo,Po Pobre"|
|Central Air|Aire acondicionado centralizado|"N No Y Sí"|
|Full Bath|Número de baños completo en planta||
|Half Bath|Número de aseos en planta||
|Bedroom AbvGr|Número de dormitorios en planta||
|Kitchen AbvGr|Número de cocinas en planta||
|Kitchen Qual|Calidad de cocinas|"Ex Excelente,Gd Bueno,TA Media,Fa Justo,Po Pobre"|
|TotRms AbvGrd|Número total de habitaciones excluidos los cuartos de baño||
|Garage Cars|Número de plazas de garaje||
|Garage Area|Superficie del garaje|||
|Garage Cond|Estado del garaje|"Ex Excelente,Gd Bueno,TA Media,Fa Justo,Po Pobre,Sg sin garaje"|
|Pool Area|Superficie de la piscina en pies cuadrados|
|Pool QC|Calidad de la piscina|"Ex Excelente,Gd Bueno,TA Media,Fa Justo,Sp no hay piscina"|
|Mo Sold|mes de venta||
|Yr Sold|año de venta||
|SalePrice|precio de venta en dólares||

Recomiendo al leer los datos, eliminar espacios de los nombres de las columnas, realiza un pequeño análisis inicial de los mismos. No olvides fijarte en los tipos de variables, que variables pueden tener tipos confundidos y corrige los. Sobre todo, trabaja con las fechas.

**NOTA:** Las tareas complementarias sirven para subir nota. El resto de preguntas valen igual y suman 10 puntos.

## 1. Inicializar y cargar el contexto spark

In [ ]:
!apt-get install openjdk-8-jdk-headless -qq > /dev/null
!wget -q https://www-us.apache.org/dist/spark/spark-2.4.7/spark-2.4.7-bin-hadoop2.7.tgz
!tar xf spark-2.4.7-bin-hadoop2.7.tgz #d
!pip install -q findspark 

import os
os.environ["SPARK_HOME"] = "/content/spark-2.4.7-bin-hadoop2.7"

import findspark
findspark.init()

from pyspark import SparkContext
sc = SparkContext()

from pyspark.sql import SparkSession
spark = SparkSession.builder.master("local[*]").getOrCreate()

In [ ]:
from pyspark.sql import SQLContext
from pyspark.sql import Row

In [ ]:
sqlContext = SQLContext(sc)

In [ ]:
final_raw = sc.textFile("BDpracticafinalCSV.csv")
final_raw.take(5)

['Order;MS Zoning;Lot Frontage;Lot Area;Land Contour;Land Slope;Overall Qual;Year Built;Year Remod/Add;Mas Vnr Type;Exter Qual;Bsmt Cond;Total Bsmt SF;Heating QC;Central Air;Full Bath;Half Bath;Bedroom AbvGr;Kitchen AbvGr;Kitchen Qual;TotRms AbvGrd;Garage Cars;Garage Area;Garage Cond;Pool Area;Pool QC;Mo Sold;Yr Sold;SalePrice',
 '1;RL;141;31770;Lvl;Gtl;6;1960;1960;Stone;TA;Gd;1080;Fa;Y;1;0;3;1;TA;7;2;528;TA;0;Sp;5;2010;215000',
 '2;RH;80;11622;Lvl;Gtl;5;1961;1961;None;TA;TA;882;TA;Y;1;0;2;1;TA;5;1;730;TA;0;Sp;6;2010;105000',
 '3;RL;81;14267;Lvl;Gtl;6;1958;1958;BrkFace;TA;TA;1329;TA;Y;1;1;3;1;Gd;6;1;312;TA;0;Sp;6;2010;172000',
 '4;RL;93;11160;Lvl;Gtl;7;1968;1968;None;Gd;TA;2110;Ex;Y;2;1;3;1;Ex;8;2;522;TA;0;Sp;4;2010;244000']

Parseamos los datos. Por una parte, los datos estáns eparados por ";". Por otra, hay que tener en cuenta que el header está en la primera fila. Es por ello que lo eliminamos y metemos los titulos a mano (Puesto que no son tantas variables), aprovechando también para deshacernos de los espacios en blanco.

In [ ]:
final_raw = final_raw.map(lambda x:x.split(";"))
header = final_raw.first()
final_raw = final_raw.filter(lambda x : x != header)

In [ ]:
final_raw2 = final_raw.map(lambda p: Row(
    Order = p[0],
    MS_Zoning = p[1],
    Lot_Frontage = p[2],
    Lot_Area = p[3],
    Land_Contour = p[4],
    Land_Slope = p[5],
    Overall_Qual = p[6],
    Year_Built = p[7],
    Year_R_A = p[8],
    Mas_Vnr_Type = p[9],
    Exter_Qual = p[10],
    Bsmt_Cond = p[11],
    Total_Bsmt_SF = p[12],
    Heating_QC = p[13],
    Central_Air = p[14],
    Full_Bath = p[15],
    Half_Bath = p[16],
    Bedroom_AbvGr = p[17],
    Kitchen_AbvGr = p[18],
    Kitchen_Qual = p[19],
    TotRms_AbvGrd = p[20],
    Garage_Cars = p[21],
    Garage_Area = p[22],
    Garage_Cond = p[23],
    Pool_Area = p[24],
    Pool_QC = p[25],
    Mo_Sold = p[26],
    Yr_Sold = p[27],
    SalePrice = p[28]
    )
)

Pasarmos a df para operar con pyspark.sql, puesto que me resulta más cómodo.

In [ ]:
final = sqlContext.createDataFrame(final_raw2)
final.registerTempTable("final")

In [ ]:
sqlContext.sql("""
    SELECT * 
    FROM final limit 10
""").show()

+-------------+---------+-----------+----------+---------+-----------+-----------+-----------+---------+----------+-------------+------------+------------+----------+--------+------------+---------+------------+-------+-----+------------+---------+-------+---------+-------------+-------------+----------+--------+-------+
|Bedroom_AbvGr|Bsmt_Cond|Central_Air|Exter_Qual|Full_Bath|Garage_Area|Garage_Cars|Garage_Cond|Half_Bath|Heating_QC|Kitchen_AbvGr|Kitchen_Qual|Land_Contour|Land_Slope|Lot_Area|Lot_Frontage|MS_Zoning|Mas_Vnr_Type|Mo_Sold|Order|Overall_Qual|Pool_Area|Pool_QC|SalePrice|TotRms_AbvGrd|Total_Bsmt_SF|Year_Built|Year_R_A|Yr_Sold|
+-------------+---------+-----------+----------+---------+-----------+-----------+-----------+---------+----------+-------------+------------+------------+----------+--------+------------+---------+------------+-------+-----+------------+---------+-------+---------+-------------+-------------+----------+--------+-------+
|            3|       Gd|      

Como se puede apreciar, los datos se han cargado perfectmente.

## 2. Cuántas viviendas distintas encontramos en el dataset? ¿Se repite alguna? Tiene sentido que haya duplicadas? ¿Qué podemos hacer con las duplicadas?

Vamos a ver cuantas viviendas podrían estar "duplicadas", viendo como se repiten los números de identificación. Representamos esos posibles duplicados.

In [ ]:
sqlContext.sql("""
select order, count(*) as num_repeticiones 
from final
group by order
having count(*)>1
""").show()

+-----+----------------+
|order|num_repeticiones|
+-----+----------------+
| 2898|               2|
| 2930|               2|
| 2929|               2|
| 2900|               2|
| 2899|               2|
| 2901|               2|
+-----+----------------+



In [ ]:
sqlContext.sql("""
select *
from final
where order = '2898' or order = '2930' or order = '2929' or order = '2900'
or order = '2899' or order = '2901'
order by order
""").show(12)

+-------------+---------+-----------+----------+---------+-----------+-----------+-----------+---------+----------+-------------+------------+------------+----------+--------+------------+---------+------------+-------+-----+------------+---------+-------+---------+-------------+-------------+----------+--------+-------+
|Bedroom_AbvGr|Bsmt_Cond|Central_Air|Exter_Qual|Full_Bath|Garage_Area|Garage_Cars|Garage_Cond|Half_Bath|Heating_QC|Kitchen_AbvGr|Kitchen_Qual|Land_Contour|Land_Slope|Lot_Area|Lot_Frontage|MS_Zoning|Mas_Vnr_Type|Mo_Sold|Order|Overall_Qual|Pool_Area|Pool_QC|SalePrice|TotRms_AbvGrd|Total_Bsmt_SF|Year_Built|Year_R_A|Yr_Sold|
+-------------+---------+-----------+----------+---------+-----------+-----------+-----------+---------+----------+-------------+------------+------------+----------+--------+------------+---------+------------+-------+-----+------------+---------+-------+---------+-------------+-------------+----------+--------+-------+
|            4|       TA|      

Así a priori podemos ver que los duplicados son prácticamente identicos. El eliminar o mantener estos duplicados dependen del sentido que le demos a la variable "Order".

Si se tratase de un código individual de identificación para cada vivienda, lo suyo sería borrarlo, puesto que supondría que se ha introducido el mismo registro dos veces en el sistema. 

Sin embargo, cabe mencionar que podría hacer referencia a lotes de viviendas. Es decir, cada identificador serviría para identificar, valga la redundancia, el bloque o urbanización al que un conjunto de viviendas pertenecen. Siendo algo común en este sector de actividad, explicaría las características y precios idénticos.

Por ello, voy a considerar que estaríamos ante este segundo supuesto, y es por ello que no voy a borrar los duplicados.

## 3. ¿Podrías decirme el total de inmuebles y el precio medio (Sale Price) de cada zona (MS Zoning)?

Agrupo los datos por zonas y calculo, para cada una, tanto el total de inmuebles en venta como el precio medio del total de estos.

In [ ]:
sqlContext.sql("""
select MS_Zoning as Zona, count(Order) as Inmuebles, avg(SalePrice) as PrecioMedio
from final
where MS_Zoning = 'RH' or MS_Zoning = 'RL' or MS_Zoning = 'RM'
group by MS_Zoning
""").show()

+----+---------+------------------+
|Zona|Inmuebles|       PrecioMedio|
+----+---------+------------------+
|  RH|       27|136419.77777777778|
|  RL|     2271|191131.80581241744|
|  RM|      462|126781.39393939394|
+----+---------+------------------+



## 4. Media de Total Bsmt SF por cada década de construcción calculada a partir de Year Built. 
## ¿Cuál es la decada de construcción con viviendas mejor acondicionadas para el frío (Heating QC)?

En primer lugar vamos a ver cuáles son los años de construcción más antiguo y reciente, para así saber cuántas décadas tenemos que agrupar.

In [ ]:
sqlContext.sql("""
select min(Year_Built), max(Year_Built)
from final
where MS_Zoning = 'RH' or MS_Zoning = 'RL' or MS_Zoning = 'RM'
""").show(20)

+---------------+---------------+
|min(Year_Built)|max(Year_Built)|
+---------------+---------------+
|           1872|           2010|
+---------------+---------------+



Como podemos apreciar, el rango de antiguedad abarca desde 1872 hasta 2010. Por ello, utilizaremos décadas desde los 1870s hasta los 2000s.

Calculámos la superficie media de los sótanos construidos en cada década. Los ordenamos por década, de más remota a más actual.

In [ ]:
sqlContext.sql("""
select avg(Total_Bsmt_SF) as Superficie_Media,
case 
when Year_Built between 1870 and 1879 then '1870s' 
when Year_Built between 1880 and 1889 then '1880s' 
when Year_Built between 1890 and 1899 then '1890s' 
when Year_Built between 1900 and 1909 then '1900s' 
when Year_Built between 1910 and 1919 then '1910s' 
when Year_Built between 1920 and 1929 then '1920s' 
when Year_Built between 1930 and 1939 then '1930s' 
when Year_Built between 1940 and 1949 then '1940s' 
when Year_Built between 1950 and 1959 then '1950s'
when Year_Built between 1960 and 1969 then '1960s' 
when Year_Built between 1970 and 1979 then '1970s' 
when Year_Built between 1980 and 1989 then '1980s' 
when Year_Built between 1990 and 1999 then '1990s' 
else '2000s' 
end as Decada
from final
where MS_Zoning = 'RH' or MS_Zoning = 'RL' or MS_Zoning = 'RM'
group by Decada
order by Superficie_Media desc
""").show()

+------------------+------+
|  Superficie_Media|Decada|
+------------------+------+
|1347.7401812688822| 2000s|
|1138.1054313099041| 1990s|
|            1086.0| 1980s|
| 1082.672268907563| 1960s|
| 968.2955223880597| 1950s|
|  953.917808219178| 1970s|
| 885.5833333333334| 1890s|
|             836.0| 1880s|
| 833.9368421052632| 1920s|
| 788.3009708737864| 1910s|
| 769.9158878504672| 1930s|
|             761.0| 1870s|
| 720.1388888888889| 1900s|
| 706.9261744966443| 1940s|
+------------------+------+



Ahora hacemos un recuento de viviendas mejor acondicionadas en cada decada. Nos basaremos en aquellas que tengan una calificación Excelente o 'Ex'. Puesto que nos interesa saber cuál es la de mejor calificación, ordenamos los datos en sentido descendente, eligiendo el primero.

In [ ]:
sqlContext.sql("""
select count(*) as Recuento,
case 
when Year_Built between 1870 and 1879 then '1870s' 
when Year_Built between 1880 and 1889 then '1880s' 
when Year_Built between 1890 and 1899 then '1890s' 
when Year_Built between 1900 and 1909 then '1900s' 
when Year_Built between 1910 and 1919 then '1910s' 
when Year_Built between 1920 and 1929 then '1920s' 
when Year_Built between 1930 and 1939 then '1930s' 
when Year_Built between 1940 and 1949 then '1940s' 
when Year_Built between 1950 and 1959 then '1950s'
when Year_Built between 1960 and 1969 then '1960s' 
when Year_Built between 1970 and 1979 then '1970s' 
when Year_Built between 1980 and 1989 then '1980s' 
when Year_Built between 1990 and 1999 then '1990s' 
else '2000s' 
end as Decada
from final 
where Heating_QC = 'Ex' and (MS_Zoning = 'RH' or MS_Zoning = 'RL' or MS_Zoning = 'RM')
group by Decada
order by Recuento desc
""").show()

+--------+------+
|Recuento|Decada|
+--------+------+
|     637| 2000s|
|     233| 1990s|
|      98| 1950s|
|      88| 1960s|
|      68| 1970s|
|      56| 1940s|
|      51| 1920s|
|      40| 1930s|
|      38| 1910s|
|      28| 1980s|
|      12| 1900s|
|       4| 1890s|
|       2| 1880s|
+--------+------+



Como se puede apreciar, la década de los 2000 es la que tiene más viviendas de mejor calidad, llegando a casi duplicar a la segunda (los 90).

### 5. ¿Cuáles son las 10 viviendas que se vendieron por un precio más elevado por metro cuadrado en el año 2009?

Seleccionamos cada vivienda con su ratio precio/superficie, y lo ordenamos de mayor a menor, seleccionando las 10 primeras. Hay que convertir de pies a metros cuadrado.

In [ ]:
caras_2009 = sqlContext.sql("""
select Order as Casa, (SalePrice/(Lot_Area/3.2808)) as Pm2
from final 
where Yr_Sold = '2009' and (MS_Zoning = 'RH' or MS_Zoning = 'RL' or MS_Zoning = 'RM')
order by Pm2 desc
""")

In [ ]:
caras_2009.show(10)

+----+------------------+
|Casa|               Pm2|
+----+------------------+
| 936| 312.9378461538462|
| 934|302.06290322580645|
| 464|250.25502645502647|
| 935|  246.263523573201|
| 408| 233.3664285714286|
| 407|230.43714285714287|
| 405|218.72000000000003|
| 933|217.71032890940566|
| 403| 216.7671428571429|
| 411|213.68368421052634|
+----+------------------+
only showing top 10 rows



## 6. Media anual por zonas del precio de venta y metros cuadrados.

En este caso he seleccionado el precio medio y la superficie media de cada casa para cada zona y cada año. Es decir, habrá un registro de cada zona en el que se recogeran estos dos datos en cada uno de los años en los que haya habido casas en el mercado. 

Ejemplo: Para la zona A, ha habido x años con casas en venta. Recogemos tanto el precio medio como superficie media del conjunto de estas casas en cada año.

In [ ]:
sqlContext.sql("""
select Yr_Sold as Year, MS_Zoning as Zona , avg(SalePrice) AS PrecioMedio, avg(Lot_Area) as TamanoMedio
from final
where MS_Zoning = 'RH' or MS_Zoning = 'RL' or MS_Zoning = 'RM'
group by Zona, Year
""").show()

+----+----+------------------+------------------+
|Year|Zona|       PrecioMedio|       TamanoMedio|
+----+----+------------------+------------------+
|2007|  RM|128947.62105263158| 6732.273684210526|
|2010|  RL|  184978.418972332|10886.185770750988|
|2009|  RL|190455.55086372362|10787.424184261037|
|2008|  RH|150333.33333333334| 8246.666666666666|
|2009|  RH|          116206.8|            6965.8|
|2007|  RL|193399.98014440434|11406.283393501804|
|2006|  RL|193654.14845360824|11224.771134020619|
|2010|  RH|          126280.0|            8105.6|
|2008|  RM| 133336.1219512195| 6713.121951219512|
|2006|  RH|140316.66666666666|            7099.5|
|2010|  RM|119681.70689655172| 6308.775862068966|
|2007|  RH|          147250.0|           7010.75|
|2008|  RL|189885.58515283844|11276.851528384279|
|2006|  RM|127502.63725490196| 6189.401960784314|
|2009|  RM|118759.86904761905| 5947.666666666667|
+----+----+------------------+------------------+



## 7. ¿Podrías decirme el total de recaudación de las casas de revistimiento (Mas Vnr Type) de piedra con respecto a las de ladrillo? ¿Hay diferencia significativa?

Recogemos la recaudación total en función del revestimiento. Por una parte, de ladrillo visto y de ladrillo normal, y por otra, de piedra.

In [ ]:
sqlContext.sql("""
select sum(SalePrice) as Recaudacion_total
from final
where Mas_Vnr_Type = 'Stone' 
""").show()

+-----------------+
|Recaudacion_total|
+-----------------+
|      6.5196277E7|
+-----------------+



In [ ]:
sqlContext.sql("""
select sum(SalePrice) as Recaudacion_total
from final
where Mas_Vnr_Type = 'BrkCmn' or Mas_Vnr_Type = 'BrkFace' 
""").show()

+-----------------+
|Recaudacion_total|
+-----------------+
|     1.89508236E8|
+-----------------+



Se aprecia que la recaudación con los revestimientos de ladrillo ha sido considerabdlemente superior, más del triple de hecho.

## 8. ¿Cuánto son más caras las viviendas con 2 cocinas, con 2 o más plazas de garaje que las que tienen 1 cocina y 1 plaza de garaje? Comparar medias y cuartiles de ambos casos

Comparamos el precio medio de ambos tipos de vivienda.

In [ ]:
viviendas1 = sqlContext.sql("""
select avg(SalePrice)
from final
where Garage_Cars >= 2 and Kitchen_AbvGr = 2 
""")

In [ ]:
viviendas2 = viviendas1 = sqlContext.sql("""
select avg(SalePrice)
from final
where Garage_Cars = 1 and Kitchen_AbvGr = 1
""").show()

+------------------------------+
|avg(CAST(SalePrice AS DOUBLE))|
+------------------------------+
|            127367.04161248375|
+------------------------------+



In [ ]:
144995.264 - 127367.042

17628.221999999994

El precio medio del primer tipo de vivienda, con múltiples plazas de garaje y dos cocinas, es 17,628.222 euros mayor.